### PR TITLE
User Impersonation Implementation 

### PR DESCRIPTION
**Pull request by shawn-hurley**
**Describe what this PR does and why we need it**:
This PR will allow us to check for users ability to run the APB, will run APB in transient name space that will hide the escalated permissions. 
Changes proposed in this pull request
 - Add open shift client and create/delete project, subject rules review ability 
 - Add copied security/validation code from origin to check if rules cover the cluster role's rules . 
 - Add ability to pull out originating user header info.

**Does this PR depend on another PR (Use this to track when PRs should be merged)**
depends-on <PR>
https://github.com/kubernetes-incubator/service-catalog/pull/1162

### Note:
This PR will depend on using openshift 3.7

### CODE REVIEW

**Comment by shawn-hurley** [comment]
[(5) Investigate 'Enable user impersonation for brokers targetting k8s' issue for 3.7 [asb]](https://trello.com/c/fiOSsq1Q/123-5-investigate-enable-user-impersonation-for-brokers-targetting-k8s-issue-for-37-asb)

**Comment by shawn-hurley** [comment]
[(5) Add Originating Identity Header to the broker [asb]](https://trello.com/c/x0P6lyk8/450-5-add-originating-identity-header-to-the-broker-asb)

**Comment by shawn-hurley** [comment]
[(13) Service Account privileges for APBs and the Broker](https://trello.com/c/KO5c6Ixp/345-13-service-account-privileges-for-apbs-and-the-broker)

**Comment by shawn-hurley** [comment]
### To Test

`curl -s -k -XPUT -u admin:admin -H "Content-Type: application/json" -H "X-Broker-API-Originating-Identity: kubernetes eyJ1c2VybmFtZSI6ICJhZG1pbiIsICJ1aWQiOiAiIn0=" -d '{"service_id":"4b058efd9ed2414e6cfdee4bda167e9a","plan_id":"default","organization_guid":"8a7d0e88-94a7-11e7-8039-c85b76064d75","space_guid":"8a7d0e88-94a7-11e7-8039-c85b76064d75","parameters":{"postgresql_database":"admin","postgresql_password":"admin","postgresql_user":"admin"},"context":{"namespace":"demo-app","platform":"kubernetes"}}' https://asb-1338-ansible-service-broker.172.17.0.1.nip.io/v2/service_instances/97f2b6c3-45ba-4e6c-a372-77c43fa0fc9e?accepts_incomplete=true`

Will use the admin user for the user info.

You will also need to specifically give asb access to create and delete. This will be added to the template in this PR. but example of the cluster role is
```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
# "namespace" omitted since ClusterRoles are not namespaced
  name: project-creator
rules:
- apiGroups: ["project.openshift.io"]
  resources: ["projects"]
  verbs: ["create", "delete"]
- apiGroups: ["authorization.openshift.io"]
  resources: ["subjectrulesreview"]
  verbs: ["create"]

```
and cluster role binding 
```yaml
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: create-projects-asb
subjects:
- kind: ServiceAccount 
  name: asb
  namespace: ansible-service-broker
roleRef:
  kind: ClusterRole
  name: project-creator 
  apiGroup: rbac.authorization.k8s.io
```

**Comment by shawn-hurley** [comment]
This will need to be run with 3.7. It is the only way that this will work.

**Comment by enj** [comment]
@openshift/sig-security 

It might be a bit before I have a chance to look at this.

**Comment by shawn-hurley** [comment]
@enj The reason for the v1 `apiVersion` is that it works for both 3.6 and 3.7. If you have a better way to do it and keep compatibility with 3.6 It would be nice to know and we can update all of the v1 references in the template.

**Comment by liggitt** [comment]
> The reason for the v1 apiVersion is that it works for both 3.6 and 3.7. If you have a better way to do it and keep compatibility with 3.6 It would be nice to know and we can update all of the v1 references in the template.

At the very least, use the grouped version of openshift APIs, e.g. `authorization.openshift.io/v1`

**Comment by shawn-hurley** [comment]
@liggitt @enj I have updated the template to use the grouped version of the openshift APIs. @jwmatthews this still works with the 3.6 version of `catasb`

**Comment by shawn-hurley** [comment]
Added #429 to capture moving all of the template yaml to RBAC API versions when we make the cut to 3.7.

File: pkg/clients/openshift.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+		Name(name).
+		Do().
+		Error()
+}
+
+// SubjectRulesReview - create and run a OpenShift Subject Rules Review
+func (o OpenshiftClient) SubjectRulesReview(user, namespace string, log *logging.Logger) (result *authorization.SubjectRulesReview, err error) {
+	body := &SubjectRulesReview{
+		Spec: SubjectRulesReviewSpec{
+			User: "admin",
```
Comment (shawn-hurley) [inline pkg/clients/openshift.go]: change to user

**Comment by jwmatthews** [comment]
Service Catalog 0.0.21 will send the Originating User header to us.

Full end to end testing of this PR is waiting for below to merge into origin:
https://github.com/openshift/origin/pull/16457

File: pkg/apb/executor.go [hunk 2]
```diff
@@ -86,18 +88,18 @@ func ExecuteApb(
 		// of uniquenes and will meet DNS name requirements.
 		executionContext.Namespace = uuid.New()
 		executionContext.Targets = append(executionContext.Targets, context.Namespace)
-		//Creating project
-		_, err := openshitftClient.CreateProject(executionContext.Namespace)
+		// Create namespace.
+		namespace := v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{transientNameSpaceKey: spec.FQName},
```
Comment (shawn-hurley) [inline pkg/apb/executor.go]: @enj creating namespaces and adding the metadata about which APB started the namespace.

File: pkg/app/app.go [hunk 3]
```diff
@@ -306,10 +307,9 @@ func retrieveClusterRoleRules(clusterRole string, log *logging.Logger) ([]author
 	if err != nil {
 		return nil, err
 	}
-	cRole := &authorization.ClusterRole{}
-	err = authorization.ConvertRBACClusterRoleToAuthorizationClusterRole(k8sRole, cRole, nil)
-	if err != nil {
+	rbacClusterRole := rbac.ClusterRole{}
+	if v1beta1rbac.Convert_v1beta1_ClusterRole_To_rbac_ClusterRole(k8sRole, &rbacClusterRole, nil); err != nil {
```
Comment (shawn-hurley) [inline pkg/app/app.go]: @enj Getting v1beta1 RBAC rules to RBAC rules.

File: pkg/clients/openshift.go [hunk 4]
```diff
@@ -171,60 +260,31 @@ func setConfigDefaultsAuth(config *rest.Config) error {
...
-		},
+	//Need to take the v1 Policy Rule and make it a Authorization Rule.
+	pr, err := autoConvertV1PolicyRulesToAuthorizationPolicyRules(r.Status.Rules)
+	if err != nil {
+		return nil, err
 	}
-	result.Kind = r.Kind
-	result.APIVersion = r.APIVersion
-	return
+	return authorization.ConvertAPIPolicyRulesToRBACPolicyRules(pr), nil
```
Comment (shawn-hurley) [inline pkg/clients/openshift.go]: @enj Converting API Policy Rules to RBAC Policy rules

File: pkg/handler/handler.go [hunk 5]
```diff
@@ -625,11 +626,11 @@ func (h handler) validateUser(userName, namespace string) (bool, int, error) {
 		return false, http.StatusInternalServerError, fmt.Errorf("Unable to connect to the cluster")
 	}
 	// Retrieving the rules for the user in the namespace.
-	res, err := openshiftClient.SubjectRulesReview(userName, namespace, h.log)
+	prs, err := openshiftClient.SubjectRulesReview(userName, namespace, h.log)
 	if err != nil {
 		return false, http.StatusInternalServerError, fmt.Errorf("Unable to connect to the cluster")
 	}
-	if covered, _ := authorization.Covers(res.Status.Rules, h.clusterRoleRules); !covered {
+	if covered, _ := validation.Covers(prs, h.clusterRoleRules); !covered {
```
Comment (shawn-hurley) [inline pkg/handler/handler.go]: @enj Using k8s RBAC [validation](https://github.com/kubernetes/kubernetes/tree/master/pkg/registry/rbac/validation)

File: templates/deploy-ansible-service-broker.template.yaml [hunk 6]
```diff
@@ -32,10 +32,10 @@ objects:
   apiVersion: authorization.openshift.io/v1
   metadata:
     # "namespace" omitted since ClusterRoles are not namespaced
-    name: project-creator
+    name: asb-auth
```
Comment (shawn-hurley) [inline templates/deploy-ansible-service-broker.template.yaml]: @enj Renamed role to something that makes a little more sense

**Comment by enj** [comment]
LGTM

There are some followup items but this is in line with our discussions.

**Comment by shawn-hurley** [comment]
I have tested with the latest service catalog and it seems to be working as expected.

File: templates/deploy-ansible-service-broker.template.yaml [hunk 7]
```diff
@@ -27,7 +27,7 @@ objects:
   metadata:
     name: asb
     namespace: ansible-service-broker
-
+
```
Comment (eriknelson) [inline templates/deploy-ansible-service-broker.template.yaml]: Can we strip trailing whitespace in this file?

File: pkg/clients/openshift.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+		out = nil
+	}
+	return out, nil
+}
+
+/* End of v1 openshift rest calls that are need. */
+
+// Openshift - Create a new openshift client if needed, returns reference
+func Openshift(log *logging.Logger) (*OpenshiftClient, error) {
+	errMsg := "Something went wrong while initializing kubernetes client!\n"
```
Comment (eriknelson) [inline pkg/clients/openshift.go]: s/kubernetes/openshift/

File: pkg/clients/openshift.go [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+			// NOTE: Looking to leverage panic recovery to gracefully handle this
+			// with things like retries or better intelligence, but the environment
+			// is probably in a unrecoverable state as far as the broker is concerned,
+			// and demands the attention of an operator.
+			panic(err.Error())
+		}
+		instances.Openshift = client
+	})
+	if instances.Openshift == nil {
+		return nil, errors.New("Kubernetes client instance is nil")
```
Comment (eriknelson) [inline pkg/clients/openshift.go]: s/Kubernetes/Openshift

File: pkg/clients/openshift.go [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+/* Start of V1 Authorizaiont rules need for openshift rest call */
+var oldAllowAllPolicyRule = PolicyRule{APIGroups: nil, Verbs: []string{"*"}, Resources: []string{"*"}}
+
+// OpenshiftClient - Client to interact with openshift api
+type OpenshiftClient struct {
+	restClient rest.Interface
+}
+
+// SubjectRulesReview is a resource you can create to determine which actions another user can perform in a namespace
+type SubjectRulesReview struct {
```
Comment (rthallisey) [inline pkg/clients/openshift.go]: Would prefer all the types in a types.go

Comment (rthallisey) [reply pkg/clients/openshift.go]: We'll move away from types.go. Ignore this comment

**Pull request merged**